# Capstone Project - Most Popular Venues in <font color=lightblue>Downtown Toronto</font> VS. <font color=Pink>Downtown Vancouver</font>

### Applied Data Science Capstone by IBM/Coursera

#### Importing Libraries

##  Introduction: 
Vancouver is the largest city in western Canada. As the most populous city in the British Columbia, there are many similarities between Vancouver and Toronto, the most populous city in Ontario. Small business owners/investers are interested in understanding what kind of business is most popular in these major cities of Canada. Clustering is utilized in this project in order to help understand what locations are major zones of interest and can be used as starting points for investment considerations.

This project aims to explore the __similarities and dissimilarities__ between Downtown Toronto and Downtown Vancouver by analyzing the most popular venues in the neighbourhoods.

## Data
The data in this analysis are from Foursquare location data, where the venue data are fetched around location of interest.

## Methodology
- Initial geographical data (Postal code, Neighborhood, City/Borough) scrapped from Wiki page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M and https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_V
- Obtain Latitude and Longitude coordinates
 - Toronto: from csv file https://cocl.us/Geospatial_data
 - Vancouver: use _geopy.geocoders_, _Nominatim_ to convert addresses into latitude and longitude values
- Machine Learning: Clustering with k-means
 - Generate top 10 venues for each neighborhood
 - Run k-means to cluster the neighborhood into 5 clusters
- Visualization of maps with location clusters generated using _folium_

Firstly, let's import some libraries.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
#!conda install -c conda-forge geopy --yes

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!pip install bs4
#!pip install plotly
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

#!pip install lxml
#!pip install html5lib

print('Libraries imported.')

Libraries imported.


# <font color=lightblue>Part 1: Analyzing Neighbourhoods in Downtown Toronto</font>

## Toronto Wiki Link URL

In [2]:
my_url1=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text   #Wiki page for Toronto Postal Codes

## 1. Create Soup Object from URL

In [3]:
soup1 = BeautifulSoup(my_url1,"html5lib")

In [4]:
tag_object1=soup1.title
print("tag object:",tag_object1)

tag object: <title>List of postal codes of Canada: M - Wikipedia</title>


In [5]:
table_contents=[]
table=soup1.find('table')
table

<table cellpadding="2" cellspacing="0" rules="all" style="width:100%; border-collapse:collapse; border:1px solid #ccc;">

<tbody><tr>
<td style="width:11%; vertical-align:top; color:#ccc;">
<p><b>M1A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
</p>
</td>
<td style="width:11%; vertical-align:top; color:#ccc;">
<p><b>M2A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M3A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>)</span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M4A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>)</span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M5A</b><br/><span style="font-size:85%;"><a h

In [6]:
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)
        
print(table_contents)


[{'PostalCode': 'M3A', 'Borough': 'North York', 'Neighborhood': 'Parkwoods'}, {'PostalCode': 'M4A', 'Borough': 'North York', 'Neighborhood': 'Victoria Village'}, {'PostalCode': 'M5A', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Regent Park, Harbourfront'}, {'PostalCode': 'M6A', 'Borough': 'North York', 'Neighborhood': 'Lawrence Manor, Lawrence Heights'}, {'PostalCode': 'M7A', 'Borough': "Queen's Park", 'Neighborhood': 'Ontario Provincial Government'}, {'PostalCode': 'M9A', 'Borough': 'Etobicoke', 'Neighborhood': 'Islington Avenue'}, {'PostalCode': 'M1B', 'Borough': 'Scarborough', 'Neighborhood': 'Malvern, Rouge'}, {'PostalCode': 'M3B', 'Borough': 'North York', 'Neighborhood': 'Don Mills North'}, {'PostalCode': 'M4B', 'Borough': 'East York', 'Neighborhood': 'Parkview Hill, Woodbine Gardens'}, {'PostalCode': 'M5B', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Garden District, Ryerson'}, {'PostalCode': 'M6B', 'Borough': 'North York', 'Neighborhood': 'Glencairn'}, {'PostalCode': 'M9

## Generating the table with PostalCode, Borough, and Neighborhood.

In [7]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

## Shape of the table:

In [8]:
df.shape

(103, 3)

## 2. Import Coordinates data frame from csv file and join to original table on postal code to get Latitude and Longitude columns

In [9]:
data_cds = pd.read_csv('https://cocl.us/Geospatial_data') 
data_cds.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## Dataframe with Coordinates:

In [10]:
df_final = pd.merge(df, data_cds, left_on='PostalCode', right_on='Postal Code', how='left').drop('Postal Code', axis=1)
df_final.head(20)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   
15        M5C  Downtown Toronto   
16        M6C              York   
17        M9C         Etobicoke   
18        M1E       Scarborough   
19        M4E      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                       Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                     Don Mills North  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724  
12             Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
13                                    Don Mills South  43.725900 -79.340923  
14                                   Woodbine Heights  43.695344 -79.318389  
15                                     St. James Town  43.651494 -79.375418  
16                                 Humewood-Cedarvale  43.693781 -79.428191  
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201  
18                  Guildwood, Morningside, West Hill  43.763573 -79.188711  
19                                        The Beaches  43.676357 -79.293031

## 3. Clustering and Mapping

In [11]:
pip install folium

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 94 kB 7.0 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [12]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

In [13]:
trt_dt_data = df_final[df_final['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
trt_dt_data

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M5B  Downtown Toronto   
2         M5C  Downtown Toronto   
3         M5E  Downtown Toronto   
4         M5G  Downtown Toronto   
5         M6G  Downtown Toronto   
6         M5H  Downtown Toronto   
7         M5J  Downtown Toronto   
8         M5K  Downtown Toronto   
9         M5L  Downtown Toronto   
10        M5S  Downtown Toronto   
11        M5T  Downtown Toronto   
12        M5V  Downtown Toronto   
13        M4W  Downtown Toronto   
14        M4X  Downtown Toronto   
15        M5X  Downtown Toronto   
16        M4Y  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1                            Garden District, Ryerson  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         Berczy Park  43.644771 -79.373306  
4                                  Central Bay Street  43.657952 -79.387383  
5                                            Christie  43.669542 -79.422564  
6                            Richmond, Adelaide, King  43.650571 -79.384568  
7   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
8            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
9                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
10                     University of Toronto, Harbord  43.662696 -79.400049  
11          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
12  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
13                                           Rosedale  43.679563 -79.377529  
14                        St. James Town, Cabbagetown  43.667967 -79.367675  
15             First Canadian Place, Underground city  43.648429 -79.382280  
16                               Church and Wellesley  43.665860 -79.383160

## Visualize the Toronto Neighbourhood

In [14]:
#pip install Nominatim
from geopy.geocoders import Nominatim

In [15]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

# create map of Manhattan using latitude and longitude values
map_trt = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(trt_dt_data['Latitude'], trt_dt_data['Longitude'], trt_dt_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_trt)  
    
map_trt

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


## Define Foursquare Credentials and Version

In [16]:

CLIENT_ID = 'GL3GBNB0DRIQ3Z5QJE54ZHPPKZARCKPRBAKRH2NDF0C4DLP2' # your Foursquare ID
CLIENT_SECRET = 'FS1GSVHKHW4NKIWWUYC4XQISQ0GEOBHVHCP0XR5W2PCBG3FR' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)




Your credentails:
CLIENT_ID: GL3GBNB0DRIQ3Z5QJE54ZHPPKZARCKPRBAKRH2NDF0C4DLP2
CLIENT_SECRET:FS1GSVHKHW4NKIWWUYC4XQISQ0GEOBHVHCP0XR5W2PCBG3FR


In [17]:


trt_dt_data.loc[0, 'Neighborhood']


neighborhood_latitude = trt_dt_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = trt_dt_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = trt_dt_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))




Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


## Let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters

In [18]:
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


results = requests.get(url).json()
results


{'meta': {'code': 200, 'requestId': '606f449c54086e3f73416729'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 43,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '53b8466a498e83df908c3f21',
       'name': 'Tandem Coffee',
       'location': {'address': '368 King St E',
        'crossStreet': 'at Trinity St',
        'lat': 43.65355870959944,
        'lng': -79.36180945913513,
        'labeledLatLngs': [{'label': 'display',
 

In [19]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
    
    
venues = results['response']['groups'][0]['items']
    
nearby_venues = pd.json_normalize(venues) # flatten JSON


In [20]:

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))


43 venues were returned by Foursquare.


## Explore Neighborhoods in Toronto

In [21]:


def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)



trt_venues = getNearbyVenues(names=trt_dt_data['Neighborhood'],
                                   latitudes=trt_dt_data['Latitude'],
                                   longitudes=trt_dt_data['Longitude']
                                  )


print(trt_venues.shape)
trt_venues.head()


Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley
(1087, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0           Tandem Coffee       43.653559       -79.361809   
1        Roselle Desserts       43.653447       -79.362017   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0          Coffee Shop  
1               Bakery  
2  Distribution Center  
3           Restaurant  
4                  Spa

## Analyze Each Neighborhood

In [22]:


trt_venues.groupby('Neighborhood').count()

print('There are {} uniques categories.'.format(len(trt_venues['Venue Category'].unique())))


# one hot encoding
trt_onehot = pd.get_dummies(trt_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
trt_onehot['Neighborhood'] = trt_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [trt_onehot.columns[-1]] + list(trt_onehot.columns[:-1])
trt_onehot = trt_onehot[fixed_columns]

trt_onehot.head()




There are 203 uniques categories.


Yoga Studio  Airport  Airport Food Court  Airport Lounge  Airport Service  \
0            0        0                   0               0                0   
1            0        0                   0               0                0   
2            0        0                   0               0                0   
3            0        0                   0               0                0   
4            0        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       1     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Candy Store  \
0             0              0        0     0            0   
1             0              0        0     0            0   
2             0              0        0     0            0   
3             0              0        0     0            0   
4             0              0        0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0                   0               0   
4                     0            0         

In [23]:

trt_onehot.shape



(1087, 203)

In [24]:

trt_grouped = trt_onehot.groupby('Neighborhood').mean().reset_index()
trt_grouped

trt_grouped.shape



(17, 203)

## Print each neighborhood along with the top 5 most common venues

In [25]:

num_top_venues = 5

for hood in trt_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = trt_grouped[trt_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
    
    


----Berczy Park----
            venue  freq
0     Coffee Shop  0.08
1    Cocktail Bar  0.05
2          Bakery  0.05
3        Pharmacy  0.03
4  Farmers Market  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3   Harbor / Marina  0.07
4             Plane  0.07


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1      Sandwich Place  0.06
2  Italian Restaurant  0.05
3                Café  0.05
4        Burger Joint  0.03


----Christie----
                venue  freq
0       Grocery Store  0.24
1                Café  0.18
2                Park  0.12
3          Baby Store  0.06
4  Italian Restaurant  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1  Japanese Restaurant  0.07
2     Sushi Restaurant  0.07
3           Restaurant  0.04
4  

In [26]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


## Create the new dataframe and display the top 10 venues for each neighborhood

In [27]:


num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = trt_grouped['Neighborhood']

for ind in np.arange(trt_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(trt_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar           Cheese Shop   
1       Airport Service      Airport Terminal           Coffee Shop   
2        Sandwich Place    Italian Restaurant                  Café   
3                  Café                  Park           Candy Store   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue     6th Most Common Venue 7th Most Common Venue  \
0            Restaurant            Farmers Market              Pharmacy   
1                 Plane             Boat or Ferry   Rental Car Location   
2          Burger Joint               Salad Place       Thai Restaurant   
3                  Bank                Baby Store    Athletics & Sports   
4           Yoga Studio  Mediterranean Restaurant                 Hotel   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant              Beer Bar              Nightclub  
1      Sculpture Garden              Boutique                Airport  
2       Bubble Tea Shop                  Park             Poke Place  
3            Restaurant    Italian Restaurant            Coffee Shop  
4           Men's Store  Fast Food Restaurant                Gay Bar

## Run k-means to cluster the neighborhood into 5 clusters

In [28]:
# set number of clusters
kclusters = 5

trt_grouped_clustering = trt_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(trt_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 


# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

trt_merged = trt_dt_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
trt_merged = trt_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

trt_merged.head() # check the last columns!



PostalCode           Borough               Neighborhood   Latitude  \
0        M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
1        M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
2        M5C  Downtown Toronto             St. James Town  43.651494   
3        M5E  Downtown Toronto                Berczy Park  43.644771   
4        M5G  Downtown Toronto         Central Bay Street  43.657952   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               0           Coffee Shop                   Pub   
1 -79.378937               0        Clothing Store           Coffee Shop   
2 -79.375418               0                  Café           Coffee Shop   
3 -79.373306               0           Coffee Shop                Bakery   
4 -79.387383               0           Coffee Shop        Sandwich Place   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                     Bakery                  Park        Breakfast Spot   
1  Middle Eastern Restaurant                  Café        Cosmetics Shop   
2             Cosmetics Shop          Cocktail Bar              Creperie   
3               Cocktail Bar           Cheese Shop            Restaurant   
4         Italian Restaurant                  Café          Burger Joint   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                  Café               Theater  Gym / Fitness Center   
1       Bubble Tea Shop          Burger Joint        Lingerie Store   
2    Italian Restaurant                   Gym        Farmers Market   
3        Farmers Market              Pharmacy    Seafood Restaurant   
4           Salad Place       Thai Restaurant       Bubble Tea Shop   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space  Performing Arts Venue  
1   Japanese Restaurant     Italian Restaurant  
2                  Park            Art Gallery  
3              Beer Bar              Nightclub  
4                  Park             Poke Place

## Visualize the resulting clusters

In [29]:

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(trt_merged['Latitude'], trt_merged['Longitude'], trt_merged['Neighborhood'], trt_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine the Five Clusters

In [30]:
trt_merged.loc[trt_merged['Cluster Labels'] == 0, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0        Clothing Store   
2   Downtown Toronto               0                  Café   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0                  Café   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                    Pub                         Bakery                  Park   
1            Coffee Shop      Middle Eastern Restaurant                  Café   
2            Coffee Shop                 Cosmetics Shop          Cocktail Bar   
3                 Bakery                   Cocktail Bar           Cheese Shop   
4         Sandwich Place             Italian Restaurant                  Café   
6                   Café                     Restaurant         Deli / Bodega   
7               Aquarium                          Hotel                  Café   
8                  Hotel                           Café    Italian Restaurant   
9             Restaurant                          Hotel                  Café   
11           Coffee Shop  Vegetarian / Vegan Restaurant    Mexican Restaurant   
14                  Café             Chinese Restaurant            Restaurant   
15                  Café                          Hotel                   Gym   
16      Sushi Restaurant            Japanese Restaurant            Restaurant   

    5th Most Common Venue     6th Most Common Venue 7th Most Common Venue  \
0          Breakfast Spot                      Café               Theater   
1          Cosmetics Shop           Bubble Tea Shop          Burger Joint   
2                Creperie        Italian Restaurant                   Gym   
3              Restaurant            Farmers Market              Pharmacy   
4            Burger Joint               Salad Place       Thai Restaurant   
6                     Gym            Clothing Store       Thai Restaurant   
7          Scenic Lookout                   Brewery   Sporting Goods Shop   
8             Salad Place                Restaurant    Seafood Restaurant   
9                     Gym        Italian Restaurant         Deli / Bodega   
11  Vietnamese Restaurant                      Park         Grocery Store   
14                    Pub        Italian Restaurant                Bakery   
15    Japanese Restaurant                Restaurant      Asian Restaurant   
16            Yoga Studio  Mediterranean Restaurant                 Hotel   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Gym / Fitness Center           Event Space  Performing Arts Venue  
1         Lingerie Store   Japanese Restaurant     Italian Restaurant  
2         Farmers Market                  Park            Art Gallery  
3     Seafood Restaurant              Beer Bar              Nightclub  
4        Bubble Tea Shop                  Park             Poke Place  
6         Cosmetics Shop      Sushi Restaurant           Concert Hall  
7             Restaurant    Italian Restaurant    Fried Chicken Joint  
8    Japanese Restaurant                Lounge    Sporting Goods Shop  
9     Seafood Restaurant   Japanese Restaurant    American Restaurant  
11           Gaming Cafe                   Bar                 Bakery  
14           Pizza Place                  Park                Butcher  
15           Salad Place   American Restaurant     Seafood Restaur

In [31]:
trt_merged.loc[trt_merged['Cluster Labels'] == 1, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13            Playground                 Trail         Women's Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
13          Dessert Shop     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13   Dumpling Restaurant            Donut Shop       Doner Restaurant

In [32]:
trt_merged.loc[trt_merged['Cluster Labels'] == 2, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5  Downtown Toronto               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5                  Café                  Park           Candy Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5                  Bank            Baby Store    Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5            Restaurant    Italian Restaurant            Coffee Shop

In [33]:
trt_merged.loc[trt_merged['Cluster Labels'] == 3, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12  Downtown Toronto               3        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12       Airport Service      Airport Terminal           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12                 Plane         Boat or Ferry   Rental Car Location   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12      Sculpture Garden              Boutique                Airport

In [34]:
trt_merged.loc[trt_merged['Cluster Labels'] == 4, trt_merged.columns[[1] + list(range(5, trt_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto               4                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10        Sandwich Place                   Bar    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10   Japanese Restaurant             Bookstore                Bakery   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
10           Yoga Studio              Beer Bar             Beer Store

# <font color=pink>Part 2: Analyzing Neighbourhoods in Downtown Vancouver</font>

## Vancover Wiki Page URL

In [35]:
my_url2=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_V').text   #Wiki page for Vancouver Postal Codes

## 1. Create Soup Object from URL

In [36]:
soup2 = BeautifulSoup(my_url2,"html5lib")

In [37]:
tag_object2=soup2.title
print("tag object:",tag_object2)

tag object: <title>List of postal codes of Canada: V - Wikipedia</title>


In [38]:
table_contents2=[]
table2=soup2.find('table')
table2

<table class="wikitable sortable">
<tbody><tr>
<td valign="top" width="11.1%"><b>V1A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Kimberley,_British_Columbia" title="Kimberley, British Columbia">Kimberley</a></span>
</td>
<td valign="top" width="11.1%"><b>V2A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Penticton" title="Penticton">Penticton</a></span>
</td>
<td valign="top" width="11.1%"><b>V3A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Langley,_British_Columbia_(district_municipality)" title="Langley, British Columbia (district municipality)">Langley Township</a><br/>(Langley City)</span>
</td>
<td valign="top" width="11.1%"><b>V4A</b><br/><span style="font-size: smaller; line-height: 125%;"><a href="/wiki/Surrey,_British_Columbia" title="Surrey, British Columbia">Surrey</a><br/>Southwest</span>
</td>
<td valign="top" width="11.1%"><b>V5A</b><br/><span style="font-size: smaller; line-height

In [39]:
for row in table2.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.text[:3]
        cell['City'] = (row.span.text).split('(')[0]
        table_contents2.append(cell)
        
print(table_contents2)


[{'PostalCode': 'V1A', 'City': 'Kimberley'}, {'PostalCode': 'V2A', 'City': 'Penticton'}, {'PostalCode': 'V3A', 'City': 'Langley Township'}, {'PostalCode': 'V4A', 'City': 'SurreySouthwest'}, {'PostalCode': 'V5A', 'City': 'Burnaby'}, {'PostalCode': 'V6A', 'City': 'Vancouver'}, {'PostalCode': 'V7A', 'City': 'RichmondSouth'}, {'PostalCode': 'V8A', 'City': 'Powell River'}, {'PostalCode': 'V9A', 'City': 'Victoria'}, {'PostalCode': 'V1B', 'City': 'VernonEast'}, {'PostalCode': 'V2B', 'City': 'KamloopsNorthwest'}, {'PostalCode': 'V3B', 'City': 'Port CoquitlamCentral'}, {'PostalCode': 'V4B', 'City': 'White Rock'}, {'PostalCode': 'V5B', 'City': 'Burnaby'}, {'PostalCode': 'V6B', 'City': 'Vancouver'}, {'PostalCode': 'V7B', 'City': 'Richmond'}, {'PostalCode': 'V8B', 'City': 'Squamish'}, {'PostalCode': 'V9B', 'City': 'Victoria'}, {'PostalCode': 'V1C', 'City': 'Cranbrook'}, {'PostalCode': 'V2C', 'City': 'KamloopsCentral and Southeast'}, {'PostalCode': 'V3C', 'City': 'Port CoquitlamSouth'}, {'PostalCod

In [40]:
df2=pd.DataFrame(table_contents2)
df2

PostalCode                           City
0          V1A                      Kimberley
1          V2A                      Penticton
2          V3A               Langley Township
3          V4A                SurreySouthwest
4          V5A                        Burnaby
5          V6A                      Vancouver
6          V7A                  RichmondSouth
7          V8A                   Powell River
8          V9A                       Victoria
9          V1B                     VernonEast
10         V2B              KamloopsNorthwest
11         V3B          Port CoquitlamCentral
12         V4B                     White Rock
13         V5B                        Burnaby
14         V6B                      Vancouver
15         V7B                       Richmond
16         V8B                       Squamish
17         V9B                       Victoria
18         V1C                      Cranbrook
19         V2C  KamloopsCentral and Southeast
20         V3C            Port CoquitlamSouth
21         V4C                 DeltaNortheast
22         V5C                        Burnaby
23         V6C                      Vancouver
24         V7C              RichmondNorthwest
25         V8C                        Kitimat
26         V9C                       Victoria
27         V1E                     Salmon Arm
28         V2E         KamloopsSouth and West
29         V3E                 CoquitlamNorth
30         V4E                      DeltaEast
31         V5E                        Burnaby
32         V6E                      Vancouver
33         V7E              RichmondSouthwest
34         V8E                       Whistler
35         V9E                       Victoria
36         V1G                   Dawson Creek
37         V2G                  Williams Lake
38         V3G                 AbbotsfordEast
39         V4G              DeltaEast Central
40         V5G                        Burnaby
41         V6G                      Vancouver
42         V7G               North Vancouver 
43         V8G                        Terrace
44         V9G                      Ladysmith
45         V1H                     VernonWest
46         V2H                  KamloopsNorth
47         V3H                     Port Moody
48         V5H                        Burnaby
49         V6H                      Vancouver
50         V7H               North Vancouver 
51         V9H        Campbell RiverOutskirts
52         V1J                  Fort St. John
53         V2J                        Quesnel
54         V3J                 CoquitlamNorth
55         V5J                        Burnaby
56         V6J                      Vancouver
57         V7J               North Vancouver 
58         V8J                  Prince Rupert
59         V9J    CourtenayNorthern Outskirts
60         V1K                        Merritt
61         V2K             Prince GeorgeNorth
62         V3K                 CoquitlamSouth
63         V4K                 DeltaNorthwest
64         V5K                      Vancouver
65         V6K                      Vancouver
66         V7K               North Vancouver 
67         V8K             Salt Spring Island
68         V9K                 Qualicum Beach
69         V1L                         Nelson
70         V2L      Prince GeorgeEast Central
71         V3L       New WestminsterNortheast
72         V4L                 DeltaSoutheast
73         V5L                      Vancouver
74         V6L                      Vancouver
75         V7L               North Vancouver 
76         V8L                         Sidney
77         V9L                         Duncan
78         V1M          Langley TownshipNorth
79         V2M      Prince GeorgeWest Central
80         V3M       New WestminsterSouthwest
81         V4M                 DeltaSouthwest
82         V5M                      Vancouver
83         V6M                      Vancouver
84         V7M               North Vancouver 
85         V8M                Central Saanich
86

In [41]:
df_van = df2.loc[df2['City'] == 'Vancouver'] #only choose the postal codes for Vancouver
df_van = df_van.head(20)
df_van

PostalCode       City
5          V6A  Vancouver
14         V6B  Vancouver
23         V6C  Vancouver
32         V6E  Vancouver
41         V6G  Vancouver
49         V6H  Vancouver
56         V6J  Vancouver
64         V5K  Vancouver
65         V6K  Vancouver
73         V5L  Vancouver
74         V6L  Vancouver
82         V5M  Vancouver
83         V6M  Vancouver
91         V5N  Vancouver
92         V6N  Vancouver
99         V5P  Vancouver
100        V6P  Vancouver
108        V5R  Vancouver
109        V6R  Vancouver
117        V5S  Vancouver

## Shape of the table:

In [42]:
df_van.shape

(20, 2)

In [43]:
pip install geocoder

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 98 kB 9.8 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


## 2. Get coordinates based on postal codes using Geopy

In [44]:
import geocoder # import geocoder
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values


## Dataframe with Coordinates:

In [45]:
lat = []
lon=[]
for i in df_van['PostalCode']:
    address = '{}'.format(i)
    geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(address)
    latitude = location.latitude
    longitude = location.longitude
    lat.append('{}'.format(latitude))
    lon.append('{}'.format(longitude))
df_van['Latitude'] = lat
df_van['Longitude'] = lon
df_van

PostalCode       City            Latitude            Longitude
5          V6A  Vancouver   49.27377017128552  -123.09947872733053
14         V6B  Vancouver   49.27811172378708  -123.11959791158985
23         V6C  Vancouver  49.287716588954346  -123.11519411838533
32         V6E  Vancouver   49.28801282544443  -123.12108056609158
41         V6G  Vancouver          49.3002702  -123.13779663860902
49         V6H  Vancouver   49.25680006013691   -123.1331282550357
56         V6J  Vancouver   49.26091372148124  -123.14577875870242
64         V5K  Vancouver   49.28171754656246   -123.0400063294856
65         V6K  Vancouver   49.26895274770836  -123.16501916876759
73         V5L  Vancouver  49.280200918758624  -123.06656328873324
74         V6L  Vancouver    49.2534450006267  -123.17202047499656
82         V5M  Vancouver   49.26007804182507   -123.0401599764032
83         V6M  Vancouver   49.23445007981276  -123.14541062858737
91         V5N  Vancouver    49.2646254249102  -123.06586883849849
92         V6N  Vancouver  49.224594327808234   -123.1686106806456
99         V5P  Vancouver   49.22275069744015  -123.06755955330607
100        V6P  Vancouver  49.218349248173894  -123.14992456434537
108        V5R  Vancouver  49.239822993458404  -123.04080030032974
109        V6R  Vancouver   49.26911809217475  -123.19719593692786
117        V5S  Vancouver   49.21605384023168  -123.04096711473133

# 3. Clustering and Mapping

In [46]:
Vancouver_dt_data = df_van.reset_index(drop=True)
Vancouver_dt_data

PostalCode       City            Latitude            Longitude
0         V6A  Vancouver   49.27377017128552  -123.09947872733053
1         V6B  Vancouver   49.27811172378708  -123.11959791158985
2         V6C  Vancouver  49.287716588954346  -123.11519411838533
3         V6E  Vancouver   49.28801282544443  -123.12108056609158
4         V6G  Vancouver          49.3002702  -123.13779663860902
5         V6H  Vancouver   49.25680006013691   -123.1331282550357
6         V6J  Vancouver   49.26091372148124  -123.14577875870242
7         V5K  Vancouver   49.28171754656246   -123.0400063294856
8         V6K  Vancouver   49.26895274770836  -123.16501916876759
9         V5L  Vancouver  49.280200918758624  -123.06656328873324
10        V6L  Vancouver    49.2534450006267  -123.17202047499656
11        V5M  Vancouver   49.26007804182507   -123.0401599764032
12        V6M  Vancouver   49.23445007981276  -123.14541062858737
13        V5N  Vancouver    49.2646254249102  -123.06586883849849
14        V6N  Vancouver  49.224594327808234   -123.1686106806456
15        V5P  Vancouver   49.22275069744015  -123.06755955330607
16        V6P  Vancouver  49.218349248173894  -123.14992456434537
17        V5R  Vancouver  49.239822993458404  -123.04080030032974
18        V6R  Vancouver   49.26911809217475  -123.19719593692786
19        V5S  Vancouver   49.21605384023168  -123.04096711473133

## Visualize the Vancouver Neighbourhood

In [47]:
address = 'Vancouver, BC'

geolocator = Nominatim(user_agent="Vancouver_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Vancouver are {}, {}.'.format(latitude, longitude))

# create map of Manhattan using latitude and longitude values
map_vcv = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Vancouver_dt_data['Latitude'], Vancouver_dt_data['Longitude'], Vancouver_dt_data['PostalCode']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_vcv)  
    
map_vcv

The geograpical coordinate of Vancouver are 49.2608724, -123.1139529.


## Explore nearby venues of downtown Vancouver

In [48]:
vcv_venues = getNearbyVenues(names=Vancouver_dt_data['PostalCode'],
                                   latitudes=Vancouver_dt_data['Latitude'],
                                   longitudes=Vancouver_dt_data['Longitude']
                                  )


print(vcv_venues.shape)
vcv_venues.head()

V6A
V6B
V6C
V6E
V6G
V6H
V6J
V5K
V6K
V5L
V6L
V5M
V6M
V5N
V6N
V5P
V6P
V5R
V6R
V5S
(548, 7)


Neighborhood Neighborhood Latitude Neighborhood Longitude  \
0          V6A     49.27377017128552    -123.09947872733053   
1          V6A     49.27377017128552    -123.09947872733053   
2          V6A     49.27377017128552    -123.09947872733053   
3          V6A     49.27377017128552    -123.09947872733053   
4          V6A     49.27377017128552    -123.09947872733053   

                                     Venue  Venue Latitude  Venue Longitude  \
0                           Creekside Park       49.274641      -123.102701   
1                                   Boxcar       49.276613      -123.100076   
2                          Pizzeria Farina       49.276636      -123.099967   
3                                 Torafuku       49.275951      -123.099814   
4  Science World at TELUS World of Science       49.273310      -123.103501   

     Venue Category  
0              Park  
1               Bar  
2       Pizza Place  
3  Asian Restaurant  
4    Science Museum

## Analyze Each Neighborhood

In [49]:
vcv_venues.groupby('Neighborhood').count()

print('There are {} uniques categories.'.format(len(vcv_venues['Venue Category'].unique())))


# one hot encoding
vcv_onehot = pd.get_dummies(vcv_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
vcv_onehot['Neighborhood'] = vcv_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [vcv_onehot.columns[-1]] + list(vcv_onehot.columns[:-1])
vcv_onehot = vcv_onehot[fixed_columns]

vcv_onehot.head(15)

There are 164 uniques categories.


Neighborhood  Accessories Store  Airport Terminal  American Restaurant  \
0           V6A                  0                 0                    0   
1           V6A                  0                 0                    0   
2           V6A                  0                 0                    0   
3           V6A                  0                 0                    0   
4           V6A                  0                 0                    0   
5           V6A                  0                 0                    0   
6           V6A                  0                 0                    0   
7           V6A                  0                 0                    0   
8           V6A                  0                 0                    0   
9           V6A                  0                 0                    0   
10          V6A                  0                 0                    0   
11          V6A                  0                 0                    0   
12          V6A                  0                 0                    0   
13          V6A                  0                 0                    0   
14          V6A                  0                 0                    0   

    Amphitheater  Aquarium  Art Gallery  Arts & Crafts Store  \
0              0         0            0                    0   
1              0         0            0                    0   
2              0         0            0                    0   
3              0         0            0                    0   
4              0         0            0                    0   
5              0         0            0                    0   
6              0         0            0                    0   
7              0         0            0                    0   
8              0         0            0                    0   
9              0         0            0                    0   
10             0         0            0                    0   
11             0         0            0                    0   
12             0         0            0                    0   
13             0         0            0                    0   
14             0         0            0                    0   

    Asian Restaurant  Athletics & Sports  Australian Restaurant  Bakery  Bank  \
0                  0                   0                      0       0     0   
1                  0                   0                      0       0     0   
2                  0                   0                      0       0     0   
3                  1                   0                      0       0     0   
4                  0                   0                      0       0     0   
5                  0                   0                      0       0     0   
6                  0                   0                      0       0     0   
7                  0                   0                      0       0     0   
8                  0                   0                      0       0     0   
9                  0                   0                      0       0     0   
10                 0                   0                      0       0     0   
11                 0                   0                      0       0     0   
12                 0                   0                      0       0     0   
13                 0                   0                      0       0     0   
14                 0                   0                      0       0     0   

    Bar  Baseball Field  Beach  Beer Garden  Boat or Ferry  Bookstore  \
0     0               0      0            0              0          0   
1     1               0      0            0              0          0   
2     0               0      0            0              0          0   
3     0               0      0            0              0          0   
4     0               0      0            0              0          0   
5     0   

In [50]:

vcv_onehot.shape



(548, 165)

In [51]:

vcv_grouped = vcv_onehot.groupby('Neighborhood').mean().reset_index()
vcv_grouped

vcv_grouped.shape



(20, 165)

## Print each neighborhood along with the top 5 most common venues

In [52]:

num_top_venues = 5

for hood in vcv_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = vcv_grouped[vcv_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
    
    


----V5K----
                          venue  freq
0  Theme Park Ride / Attraction  0.12
1                   Event Space  0.08
2                          Park  0.08
3                    Theme Park  0.08
4                   Beer Garden  0.08


----V5L----
                venue  freq
0         Coffee Shop  0.09
1  Chinese Restaurant  0.09
2              Bakery  0.09
3             Brewery  0.09
4                Café  0.06


----V5M----
                    venue  freq
0  Furniture / Home Store  0.13
1                Bus Stop  0.09
2             Coffee Shop  0.09
3             Pizza Place  0.09
4           Metro Station  0.04


----V5N----
                  venue  freq
0           Coffee Shop  0.14
1      Sushi Restaurant  0.10
2           Pizza Place  0.07
3        Sandwich Place  0.05
4  Ethiopian Restaurant  0.05


----V5P----
                       venue  freq
0          Indian Restaurant  0.05
1  Middle Eastern Restaurant  0.05
2                 Restaurant  0.05
3                Pizza P

## Create the new dataframe and display the top 10 venues for each neighborhood

In [53]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted2 = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted2['Neighborhood'] = vcv_grouped['Neighborhood']

for ind in np.arange(trt_grouped.shape[0]):
    neighborhoods_venues_sorted2.iloc[ind, 1:] = return_most_common_venues(vcv_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted2.head()

Neighborhood         1st Most Common Venue 2nd Most Common Venue  \
0          V5K  Theme Park Ride / Attraction                  Park   
1          V5L                   Coffee Shop    Chinese Restaurant   
2          V5M        Furniture / Home Store           Coffee Shop   
3          V5N                   Coffee Shop      Sushi Restaurant   
4          V5P                          Park     Convenience Store   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Theme Park           Event Space           Beer Garden   
1                    Brewery                Bakery         Deli / Bodega   
2                Pizza Place              Bus Stop            Restaurant   
3                Pizza Place        Sandwich Place  Ethiopian Restaurant   
4  Middle Eastern Restaurant           Bus Station          Liquor Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   Fair                  Farm        Sandwich Place   
1  Vietnamese Restaurant                  Café               Theater   
2            Bus Station             Pet Store         Metro Station   
3          Grocery Store    Chinese Restaurant          Burger Joint   
4         Sandwich Place       Motorcycle Shop            Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0          Soccer Field       Sushi Restaurant  
1        Breakfast Spot    Japanese Restaurant  
2    Chinese Restaurant           Liquor Store  
3         Smoothie Shop          Bowling Alley  
4      Sushi Restaurant    Fried Chicken Joint

## Run k-means to cluster the neighborhood into 5 clusters

In [54]:
# set number of clusters
kclusters = 5

vcv_grouped_clustering = vcv_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(vcv_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

# add clustering labels
neighborhoods_venues_sorted2.insert(0, 'Cluster Labels', kmeans.labels_)

#vcv_merged = Vancouver_dt_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
#vcv_merged = vcv_merged.join(neighborhoods_venues_sorted2.set_index('Neighborhood'), on='Neighborhood')

#vcv_merged.head() # check the last columns!

neighborhoods_venues_sorted2.head()



Cluster Labels Neighborhood         1st Most Common Venue  \
0               1          V5K  Theme Park Ride / Attraction   
1               2          V5L                   Coffee Shop   
2               2          V5M        Furniture / Home Store   
3               2          V5N                   Coffee Shop   
4               1          V5P                          Park   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0                  Park                 Theme Park           Event Space   
1    Chinese Restaurant                    Brewery                Bakery   
2           Coffee Shop                Pizza Place              Bus Stop   
3      Sushi Restaurant                Pizza Place        Sandwich Place   
4     Convenience Store  Middle Eastern Restaurant           Bus Station   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0           Beer Garden                   Fair                  Farm   
1         Deli / Bodega  Vietnamese Restaurant                  Café   
2            Restaurant            Bus Station             Pet Store   
3  Ethiopian Restaurant          Grocery Store    Chinese Restaurant   
4          Liquor Store         Sandwich Place       Motorcycle Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Sandwich Place          Soccer Field       Sushi Restaurant  
1               Theater        Breakfast Spot    Japanese Restaurant  
2         Metro Station    Chinese Restaurant           Liquor Store  
3          Burger Joint         Smoothie Shop          Bowling Alley  
4            Restaurant      Sushi Restaurant    Fried Chicken Joint

In [55]:
#Join two tables
vcv_merged = Vancouver_dt_data.join(neighborhoods_venues_sorted2)
vcv_merged

PostalCode       City            Latitude            Longitude  \
0         V6A  Vancouver   49.27377017128552  -123.09947872733053   
1         V6B  Vancouver   49.27811172378708  -123.11959791158985   
2         V6C  Vancouver  49.287716588954346  -123.11519411838533   
3         V6E  Vancouver   49.28801282544443  -123.12108056609158   
4         V6G  Vancouver          49.3002702  -123.13779663860902   
5         V6H  Vancouver   49.25680006013691   -123.1331282550357   
6         V6J  Vancouver   49.26091372148124  -123.14577875870242   
7         V5K  Vancouver   49.28171754656246   -123.0400063294856   
8         V6K  Vancouver   49.26895274770836  -123.16501916876759   
9         V5L  Vancouver  49.280200918758624  -123.06656328873324   
10        V6L  Vancouver    49.2534450006267  -123.17202047499656   
11        V5M  Vancouver   49.26007804182507   -123.0401599764032   
12        V6M  Vancouver   49.23445007981276  -123.14541062858737   
13        V5N  Vancouver    49.2646254249102  -123.06586883849849   
14        V6N  Vancouver  49.224594327808234   -123.1686106806456   
15        V5P  Vancouver   49.22275069744015  -123.06755955330607   
16        V6P  Vancouver  49.218349248173894  -123.14992456434537   
17        V5R  Vancouver  49.239822993458404  -123.04080030032974   
18        V6R  Vancouver   49.26911809217475  -123.19719593692786   
19        V5S  Vancouver   49.21605384023168  -123.04096711473133   

    Cluster Labels Neighborhood         1st Most Common Venue  \
0                1          V5K  Theme Park Ride / Attraction   
1                2          V5L                   Coffee Shop   
2                2          V5M        Furniture / Home Store   
3                2          V5N                   Coffee Shop   
4                1          V5P                          Park   
5                1          V5R                           Bar   
6                2          V5S            Chinese Restaurant   
7                1          V6A                          Park   
8                1          V6B                         Hotel   
9                2          V6C                 Boat or Ferry   
10               1          V6E                         Hotel   
11               1          V6G                         Trail   
12               2          V6H                Breakfast Spot   
13               2          V6J                   Coffee Shop   
14               2          V6K                          Park   
15               4          V6L            Italian Restaurant   
16               2          V6M            Chinese Restaurant   
17               0          V6N                           NaN   
18               3          V6P                           NaN   
19               1          V6R                           NaN   

           2nd Most Common Venue      3rd Most Common Venue  \
0                           Park                 Theme Park   
1             Chinese Restaurant                    Brewery   
2                    Coffee Shop                Pizza Place   
3               Sushi Restaurant                Pizza Place   
4              Convenience Store  Middle Eastern Restaurant   
5                          Hotel                   Bus Stop   
6                  Deli / Bodega                       Bank   
7                   Dessert Shop                        Bar   
8                     Restaurant        Japanese Restaurant   
9                     Restaurant                      Hotel   
10           American Restaurant                 Restaurant   
11  Theme Park Ride / Attraction                Zoo Exhibit   
12                      Bus Stop                Coffee Shop   
13                           Bar                Yoga Studio   
14                   Coffee Shop                       Café   
15          Caribbean Restaurant             Baseball Field   
16                      Bus Stop           Sushi Restaurant   
17                           NaN                     

## Visualize the resulting clusters

In [56]:

# create map
map_clusters2 = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(vcv_merged['Latitude'], vcv_merged['Longitude'], vcv_merged['Neighborhood'], vcv_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters2)
       
map_clusters2

## Examine the Five Clusters

In [57]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 0, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
17  Vancouver          V6N                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
17                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
17                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
17                   NaN                    NaN

In [58]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 1, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City Neighborhood         1st Most Common Venue  \
0   Vancouver          V5K  Theme Park Ride / Attraction   
4   Vancouver          V5P                          Park   
5   Vancouver          V5R                           Bar   
7   Vancouver          V6A                          Park   
8   Vancouver          V6B                         Hotel   
10  Vancouver          V6E                         Hotel   
11  Vancouver          V6G                         Trail   
19  Vancouver          V6R                           NaN   

           2nd Most Common Venue      3rd Most Common Venue  \
0                           Park                 Theme Park   
4              Convenience Store  Middle Eastern Restaurant   
5                          Hotel                   Bus Stop   
7                   Dessert Shop                        Bar   
8                     Restaurant        Japanese Restaurant   
10           American Restaurant                 Restaurant   
11  Theme Park Ride / Attraction                Zoo Exhibit   
19                           NaN                        NaN   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0            Event Space           Beer Garden                  Fair   
4            Bus Station          Liquor Store        Sandwich Place   
5      Fish & Chips Shop      Asian Restaurant                  Park   
7              Gastropub           Pizza Place      Community Center   
8     Italian Restaurant    Seafood Restaurant        Cosmetics Shop   
10    Miscellaneous Shop            Steakhouse          Dessert Shop   
11                Garden                  Park                   Pub   
19                   NaN                   NaN                   NaN   

   7th Most Common Venue 8th Most Common Venue       9th Most Common Venue  \
0                   Farm        Sandwich Place                Soccer Field   
4        Motorcycle Shop            Restaurant            Sushi Restaurant   
5            Zoo Exhibit     Food & Drink Shop  Financial or Legal Service   
7            Coffee Shop           Music Store              Sandwich Place   
8                    Spa           Coffee Shop                Concert Hall   
10                  Park        Ice Cream Shop                Liquor Store   
11     Outdoor Sculpture  Other Great Outdoors                 Event Space   
19                   NaN                   NaN                         NaN   

   10th Most Common Venue  
0        Sushi Restaurant  
4     Fried Chicken Joint  
5    Fast Food Restaurant  
7                  Circus  
8       French Restaurant  
10     Seafood Restaurant  
11            Music Venue  
19                    NaN

In [59]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 2, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City Neighborhood   1st Most Common Venue 2nd Most Common Venue  \
1   Vancouver          V5L             Coffee Shop    Chinese Restaurant   
2   Vancouver          V5M  Furniture / Home Store           Coffee Shop   
3   Vancouver          V5N             Coffee Shop      Sushi Restaurant   
6   Vancouver          V5S      Chinese Restaurant         Deli / Bodega   
9   Vancouver          V6C           Boat or Ferry            Restaurant   
12  Vancouver          V6H          Breakfast Spot              Bus Stop   
13  Vancouver          V6J             Coffee Shop                   Bar   
14  Vancouver          V6K                    Park           Coffee Shop   
16  Vancouver          V6M      Chinese Restaurant              Bus Stop   

   3rd Most Common Venue            4th Most Common Venue  \
1                Brewery                           Bakery   
2            Pizza Place                         Bus Stop   
3            Pizza Place                   Sandwich Place   
6                   Bank                   Farmers Market   
9                  Hotel                      Coffee Shop   
12           Coffee Shop                           Bakery   
13           Yoga Studio              Japanese Restaurant   
14                  Café  Southern / Soul Food Restaurant   
16      Sushi Restaurant                Electronics Store   

   5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
1          Deli / Bodega  Vietnamese Restaurant                  Café   
2             Restaurant            Bus Station             Pet Store   
3   Ethiopian Restaurant          Grocery Store    Chinese Restaurant   
6          Shopping Mall           Liquor Store      Sushi Restaurant   
9                   Café                  Plaza             Hotel Bar   
12        Cosmetics Shop           Gourmet Shop     Korean Restaurant   
13     Electronics Store       Greek Restaurant     Food & Drink Shop   
14                  Bank           Liquor Store            Restaurant   
16           Coffee Shop       Asian Restaurant           Zoo Exhibit   

   8th Most Common Venue   9th Most Common Venue      10th Most Common Venue  
1                Theater          Breakfast Spot         Japanese Restaurant  
2          Metro Station      Chinese Restaurant                Liquor Store  
3           Burger Joint           Smoothie Shop               Bowling Alley  
6         Sandwich Place      Salon / Barbershop                 Gas Station  
9             Food Truck                     Spa              Breakfast Spot  
12   Japanese Restaurant            Burger Joint                 Art Gallery  
13            Restaurant  Furniture / Home Store              Breakfast Spot  
14    Chinese Restaurant              Food Truck              Breakfast Spot  
16     Food & Drink Shop       Fish & Chips Shop  Financial or Legal Service

In [60]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 3, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
18  Vancouver          V6P                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
18                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
18                   NaN                    NaN

In [61]:
vcv_merged.loc[vcv_merged['Cluster Labels'] == 4, vcv_merged.columns[[1] + list(range(5, vcv_merged.shape[1]))]]

City Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
15  Vancouver          V6L    Italian Restaurant  Caribbean Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
15        Baseball Field           Zoo Exhibit                  Farm   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
15            Food Court     Food & Drink Shop     Fish & Chips Shop   

         9th Most Common Venue 10th Most Common Venue  
15  Financial or Legal Service   Fast Food Restaurant

## Venue popularity and distribution between downtown Vancouver and downtown Toronto

In [62]:
print(vcv_merged['1st Most Common Venue'].value_counts(normalize=True))
print(trt_merged['1st Most Common Venue'].value_counts(normalize=True))

Coffee Shop                     0.176471
Park                            0.176471
Hotel                           0.117647
Chinese Restaurant              0.117647
Theme Park Ride / Attraction    0.058824
Breakfast Spot                  0.058824
Bar                             0.058824
Furniture / Home Store          0.058824
Boat or Ferry                   0.058824
Trail                           0.058824
Italian Restaurant              0.058824
Name: 1st Most Common Venue, dtype: float64
Coffee Shop       0.588235
Café              0.176471
Clothing Store    0.058824
Grocery Store     0.058824
Airport Lounge    0.058824
Park              0.058824
Name: 1st Most Common Venue, dtype: float64


## Analysis Results and Conclusions: 



- Similarity - When comparing <font color=lightblue>downtown Toronto</font> with <font color=pink>downtown Vancouver</font>, the No.1 most common venue is <font color=brown>Coffee Shop</font>  for both districts (65% vs. 18%). 
- Dissimilarity - In <font color=pink>Downtown Vancouver</font>, Hotel (12%) and Pizza Place (12%) are among the most common venues other than <font color=brown>Coffee Shop</font>. In <font color=lightblue>downtown Toronto</font>, Cafe (12%) is among the most common venues other than <font color=brown>Coffee Shop</font>.

- In Toronto downtown, the largest cluster of point of interest is centered around the Financial District. 
- In Vancouver downtown, the largest clusters of point of interest are centered around west Vancouver and north Vancouver.

## Discussion/Recommendation:



When investing in businesses in major cities like Toronto or Vancouver, consider targeting Coffee Shops, since they are the most popular venues in these big cities.

Comparing Toronto and Vancouver downtown, coffee shops are more densely located in Toronto. To reduce competition, consider investing in coffee shops in downtown Vancouver, compared to downtown Toronto.

## Future Analysis Directions:

- Further increase number of data points included and cluster sizes to improve analysis accuracy and generalizability.
- Introduce more features to the clusters, such as customer gender and age group.
- Link data to other sources such as customer reviews and ratings for coffee shops in downtown Vancouver/Toronto to gain more insights into the profitibility and marketing values of this kind of business.